feat: 멤버 미션 API를 작성한다

devholic22 opened the pull request:
## 📄 Summary

재윤님이 작성해주신 멤버 미션 API 뼈대를 기반으로 작업하였습니다.

>

## 🙋🏻 More

* 기존 재윤님이 작성해주셨던 `Mission`에 대한 예외들이 `ExceptionHandler`가 없어서 반영하였습니다.
* MemberMission의 earnReward를 보니 doesGetReward를 CLEAR_STATUS로 바꾸는 것 같아, 클리어할 때 doesGetReward도 호출하도록 하였습니다. 혹시 클리어를 하는 것과 미션의 보상을 수령하는 시점이 분리되어야 한다면 수정하도록 하겠습니다.
* 클리어 여부로 조회하는 `MemberMissionsQueryService`의 `findMemberMissionsByStatus` 메서드에서는 페이징을 적용하지 않고 개발하였습니다. 페이징을 필요로 하는 것은 `findMemberMissionsWithPaging`에서만 필요한 것으로 생각하고 개발을 했었는데, 이 또한 페이징을 추가해야 한다면 반영하도록 하겠습니다. (다만, 민호 님께 여쭤보니 회원의 미션을 조회하는 API는 필요하지 않은 영역일수도 있다고 하셨습니다.)
* 쿼리 조회가 사용되는 `MemberMissionsQueryService`는 `MemberMissionRepository`의 메서드가 조회밖에 없고, 저장이 `MemberMissionsRepository`에서 진행되기 때문에 (cascade) `MemberMissionQueryRepositoryTest`를 통합 테스트 하는 방식으로 대체하였습니다.

이번에도 고쳐야 할 점이 있다면 반영하도록 하겠습니다!

* * *
## 2차 수정
* 민호님과의 회의를 거친 뒤, 기존에 작성했던 API 상당 부분을 제거하였습니다. (보상 수령, 클리어 여부에 따른 목록 조회 등)
* API는 회원의 MemberMission 목록 조회만 가능합니다.
* 유저가 특정 미션을 수행할 경우, `MemberMissionClearedEvent`가 발생된다고 가정하였습니다. 회원 id, 미션 id, 회원 성별이 필요합니다.
* 챌린지 미션은 1번만 등록될 수 있습니다.
* 데일리 미션은 성별에 따라 "같은 날" 등록될 수 있는 갯수가 다릅니다. (여자 3개, 남자 2개까지 가능) `MemberMissions`에서 Map으로 미리 넣어두도록 하였습니다.
* 데일리 미션을 검증하기 위해서는 createdAt이 null이 되면 안됩니다. 그렇기 때문에 기존 Builder를 SuperBuilder로 수정하였습니다.
* MemberMissions에 MemberMission이 등록될 때는 클리어 한 미션만 들어올 수 있기 때문에, MemberMission에서 클리어 여부에 대한 필드를 삭제하였습니다.
* 데일리 미션이 여러 개 들어올 수 있다는 것은, 여러 개의 MemberMission이 같은 Member를 의존한다는 것을 의미합니다. 그렇기 때문에 MemberMission : Mission 관계를 ManyToOne으로 수정하였습니다.
* 보상을 받는 메서드 (receiveReward)는 남겨두었습니다. 미션 (ex: 좋아요 누르기)을 수행한 application 패키지 등에서 receiveReward 등으로 사용자에게 보상을 줄 것이라 생각했고, 그러기 위해서는 receiveReward 메서드는 남겨두는 게 맞겠다는 생각이 들었습니다.

피드백 주시면 다시 반영하겠습니다!

close #19

devholic22 committed:
feat: reward 조회 기능 구현

- 특정 (완료된) 미션에 대한 reward 조회 기능 구현
- 모든 (완료된) 미션에 대한 reward 총합 조회 기능 구현
- reward 응답 DTO 작성
- MemberMissionsRepository JPA 타입 변경

devholic22 committed:
refactor: MemberMissions 예외 패키지 추가

- MemberMissions 하위 예외 패키지 추가
- 예외 핸들러 추가

devholic22 committed:
feat: 로그인 회원 인식 인터셉터 추가

- API 경로의 회원 인식되도록 인터셉터 메서드 추가

devholic22 committed:
feat: 회원 미션 목록 (MemberMissions) 등록 구현

- 회원 미션 목록 및 미션 등록 구현
- 관련 리포지터리 인터페이스 정의

devholic22 committed:
feat: 회원 미션 클리어 구현

- 회원의 특정 미션 클리어 처리 구현

devholic22 committed:
refactor: MemberMissions 생성 함수화 수정

devholic22 committed:
refactor: MemberMissionsController 줄바꿈 추가

devholic22 committed:
refactor: membermissions 경로 이름 수정

devholic22 committed:
refactor: isExistMemberMissions 삭제

- MemberMissions isExistMemberMissions 메서드 삭제

devholic22 committed:
test: MemberMissionsJpaRepositoryTest 작성

- MemberMissionsJpaRepositoryTest 작성
- 리포지터리 메서드의 반환 타입 변경

devholic22 committed:
refactor: 구현할 필요가 없는 메서드 삭제

- MemberMissionsJpaRepository의 필요 없는 메서드 정의 삭제

devholic22 committed:
test: MemberMissionsServiceTest 작성

- 서비스 단계인 MemberMissionsService 테스트 작성
- 필요한 MemberMissionsFakeRepository 생성

devholic22 committed:
refactor: MemberMissionsServiceTest 수정

- MemberMissionsService 예외 테스트 Nested 추가
- 미션 클리어 테스트 추가
- given 표현 수정

devholic22 committed:
test: MemberMissionsControllerWebMvcTest 작성

- MemberMissions 컨트롤러 테스트 작성
- MemberMissions 문서화 적용
- 필요한 RewardResponse Fixture 생성

devholic22 committed:
feat: 회원 미션 조회 구현

- 회원의 모든 미션 (페이징) 조회 구현
- 회원의 미션 완료 여부에 따른 결과 조회 구현

devholic22 committed:
test: 회원 미션 조회 테스트 작성

- 회원 미션 조회 테스트 작성
- API 문서화 추가

devholic22 committed:
fix: MemberMissionFixture 객체 속성 수정

- 완료된 MemberMission 생성 시 doesGetReward도 true로 되도록 수정

devholic22 committed:
refactor: Mission 예외 핸들러 등록

- Mission 예외 핸들러 등록
- Mission 예외 패키지 분리

devholic22 committed:
test: MemberMissions 인수 테스트 작성

- MemberMissions 인수 테스트 작성

devholic22 committed:
refactor: final 인자 추가

- MemberMissionPagingResponse final 인자 추가

devholic22 committed:
refacor: MemberMissionsService 주석 제거

- MemberMissionsService 주석 제거

devholic22 committed:
refactor: MemberMissionsController 인자 줄내림 추가

- MemberMissionsController 메서드 인자 줄내림 추가

devholic22 committed:
refactor: 메서드 순서 조정

- MemberMissionsController, MemberMissionsService, 테스트 간의 메서드 순서 같도록 수정

devholic22 committed:
test: MemberMissionQueryRepositoryTest 작성

- MemberMissionQueryRepository 테스트 작성
- MemberMissionQueryRepository 메서드 타입 변경

devholic22 committed:
docs: MemberMissions 문서화 수정

- 컨트롤러 흐름에 맞게 문서화 수정

devholic22 committed:
refactor: MemberMissionsExceptionHandler 메서드명 수정

- MemberMissionsExceptionHandler 메서드명 수정

devholic22 committed:
refactor: ignoringFields 제거 후 다른 방식으로 교체

- ignoringFields 대신 ignoringFieldsOfTypes을 사용하도록 수정

devholic22 committed:
refactor: 인터페이스 메서드 공백 추가

- 구현체와 같은 표현을 위해 메서드 공백 추가

devholic22 committed:
refactor: 테스트 클래스 public 제거

- 테스트 클래스 public 제거

eom-tae-in commented on src/main/java/com/atwoz/mission/intrastructure/membermission/MemberMissionQueryRepository.java:
```diff
@@ -0,0 +1,57 @@
+                .innerJoin(memberMissions1)
+                .on(memberMissions1.memberMissions.contains(memberMission))
+                .where(memberMissions1.memberId.eq(memberId))
+                .where(memberMission.isStatusClear.eq(isStatusClear))
```
where절 조건이 여러개인 경우에 ,(comma)로 연결해줄 수 있습니다.

예) where(memberMissions1.memberId.eq(memberId), memberMission.isStatusClear.eq(isStatusClear));
인지들을 매서드로 분리하면 더 좋을 거 같네요!

devholic22 replied:
where절 조건을 콤마로 한 번에 같이 묶을 수 있다는 것은 몰랐네요..! 반영하였습니다

eom-tae-in commented on src/main/java/com/atwoz/mission/intrastructure/membermission/dto/MemberMissionPagingResponse.java:
```diff
@@ -0,0 +1,27 @@
+        }
+
+        return NO_MORE_PAGE;
+    }
```
dto에 nextPage를 계산하는 로직이 없어도 될 것 같습니다. 
현재 querydsl을 사용하여 화면에 보여주기 위한 쿼리를 따로 작성중인데 그 작업을 dto에서도 일부 처리하는 것 같아 보입니다.
dto가 데이터를 운반하는 역할만 가지게 하는 것이 좋아보입니다!

eom-tae-in commented on src/main/java/com/atwoz/mission/application/membermission/MemberMissionsQueryService.java:
```diff
@@ -1,30 +1,29 @@
-        // TODO
-        return;
+    public List<MemberMissionSimpleResponse> findMemberMissionsByStatus(final Long memberId, final boolean isStatusClear) {
+        return memberMissionRepository.findMemberMissionsByStatus(memberId, isStatusClear);
```
MemberMissionSimpleResponse가 여러 곳에서 사용되는 것으로 파악되는데 화면 출력을 위한 dto는 따로 클래스를 만들어서 반환하는게 어떨까요?
만약 findMemberMissionsByStatus 메서드가 반환하는 데이터의 형식이 바뀌게 된다면 새로운 dto 클래스가 필요하지 않을까요?
클래스를 재사용하기 위해 MemberMissionSimpleResponse를 반환타입으로 사용하신 것 같아보이지만 출력에 해당하는 dto를 따로 생성하면 더 좋을 것 같다는 생각이 듭니다!

eom-tae-in commented on src/main/java/com/atwoz/mission/application/membermission/MemberMissionsService.java:
```diff
@@ -9,24 +15,43 @@
+        MemberMissions memberMissions = memberMissionsRepository.findByMemberId(memberId)
+                .orElseGet(() -> createNewMemberMissionsWithMemberId(memberId));
+
+        Mission mission = missionRepository.findById(missionId)
+                .orElseThrow(MissionNotFoundException::new);
```
이 코드도 메서드로 분리하는게 더 좋아보입니다~

devholic22 replied:
수정하였습니다!

eom-tae-in commented on src/main/java/com/atwoz/mission/application/membermission/MemberMissionsService.java:
```diff
@@ -9,24 +15,43 @@
 
-    public void getRewardByMissionId() {
+    public void addMemberMission(final Long memberId, final Long missionId) {
+        MemberMissions memberMissions = memberMissionsRepository.findByMemberId(memberId)
+                .orElseGet(() -> createNewMemberMissionsWithMemberId(memberId));
```
해당 로직을 메서드로 분리하는게 어떨까요?
그리고 orElseGet보다 existByMemberId로 존재하는지 확인하고 없는 경우 생성해주는 방식이 가독성이 더 좋아보입니다!

sosow0212 replied:
전 exist도 좋지만 도메인적으로 처리하기 위해서 orElseGet도 괜찮다는 입장이긴합니다

현준님이 직접 둘의 차이를 작성해보면 어떨까요? (성능뿐만 아니라 객체지향적으로 봤을 때)

devholic22 replied:
이 점이 많이 고민되는 점인 것 같습니다.

1만 건의 더미 MemberMissions를 만든 다음 orElseGet과 exists의 방식을 비교해봤는데, 어떤 때는 orElseGet이 더 빠르고 어떤 때는 exists 방식이 더 빠름을 확인했습니다. (격차는 약 200ms 정도였습니다. 테스트를 측정할 때 두 방식 모두 인텔리제이의 캐시를 삭제한 뒤 진행하는 방식으로 했는데, 혹시 더 정확하게 성능을 비교할 수 있는 방법이 있는지도 알아보겠습니다.)

그렇다면 남은 부분은 도메인적, 가독성 부분일 것 같은데 가독성에서는 exists 방식이 "데이터베이스에서 조회했을 때 존재하면 기존 것을 쓰고, 존재하지 않으면 새로 만든다" 느낌을 주므로 (리포지터리 메서드를 직접 작성하므로) 더 나을 것 같다는 생각이 들며,
도메인적으로는 orElseGet이 리포지터리 메서드 반환 타입인 Optional의 특성 (orElseGet)을 이용하고, 없을 때 어떻게 처리를 할 지 하나의 줄에서 바로 명시하므로 더 나을 것 같다는 생각이 들었습니다.

그렇지만 저희가 지향하는 설계는 아무래도 도메인 주도 개발이다보니 재윤님이 말씀하신 orElseGet 방식이 조금 더 맞을 것 같다는 생각이 들었습니다. (그리고 성능이 많이 차이나지 않는다고는 하지만, 이미 MemberMissions가 저장된 경우에는 orElseGet은 바로 select 한 번, exists 방식은 exists -> select로 쿼리가 두 번 호출되기 때문에 select를 한 번만 하는 orElseGet이 조금 더 좋을 것 같다는 생각이 듭니다.)

명확히 어떤 것이 좋다는 결론을 내기 어려운 고민 같은데 이 의견 이후의 태인님의 의견이 궁금합니다!

eom-tae-in replied:
저도 좀 더 생각해 보니 지금 같은 경우에는 orElseGet으로 처리하는 것이 좋을 거 같습니다!

eom-tae-in reviewed (changes requested):
고생하셨습니다. 몇가지만 수정해주세요~

현준님께서 남겨주신 comment를 보니까 MemberMissionsQueryService를 테스트를 진행하지 않으신거 같은데 해당 service는 테스트를 진행하는 것이 맞아보입니다. service 계층은 실제 구현에 종속되지 않아 mock, fake 객체로 테스트를 진행하기 때문에 테스트를 작성해주시는 것이 맞습니다

devholic22 committed:
refactor: where 콤마 방식으로 수정

- MemberMissionQueryRepository where 콤마 방식으로 수정

devholic22 committed:
refactor: Mission 조회 함수화

- MemberMissionsService Mission 조회 함수화

sosow0212 commented on src/main/java/com/atwoz/member/config/AuthConfig.java:
```diff
@@ -37,7 +37,8 @@ public void addInterceptors(InterceptorRegistry registry) {
                 .excludePathPattern("/**", OPTIONS)
-                .addPathPatterns("/api/info/**", ANY);
+                .addPathPatterns("/api/info/**", ANY)
+                .addPathPatterns("/api/members/me/missions/**", ANY);
```
사용하는 메서드만 열어놓는 게 좋을 것 같아요

sosow0212 commented on src/main/java/com/atwoz/mission/domain/membermission/MemberMissions.java:
```diff
@@ -38,10 +39,27 @@ public class MemberMissions extends BaseEntity {
+        MemberMission targetMemberMission = this.memberMissions.stream()
+                .filter(memberMission -> memberMission.isSameMission(missionId))
+                .findAny()
+                .orElseThrow(MemberMissionNotFoundException::new);
```
예외 메시지를 보면 미션이 아예 없다는 예외를 던지는데 해당 경우에서는 프론트에서 헷갈릴 수 있을 것 같아요
조금 더 명확하게 작성해주는게 좋을 것 같습니다

sosow0212 commented on src/main/java/com/atwoz/mission/exception/mission/MissionExceptionHandler.java:
```diff
@@ -0,0 +1,50 @@
+        return getBadRequestResponse(e);
+    }
+
+    private ResponseEntity<String> getBadRequestResponse(final Exception e) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+                .body(e.getMessage());
+    }
+
+    private ResponseEntity<String> getNotFoundResponse(final Exception e) {
+        return ResponseEntity.status(HttpStatus.NOT_FOUND)
+                .body(e.getMessage());
+    }
```
이것도 매개변수에서 HttpStatus를 같이 받는다면 줄일 수 있을 것 같습니다
그리고 e보다는 exception 같은 풀 네이밍을 적는게 좋을 것 같습니다

sosow0212 commented on src/main/java/com/atwoz/mission/domain/membermission/MemberMissionRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.List;
+
+public interface MemberMissionRepository {
```
MemberMissionRepository가 필요하나요? OneToMany로 1:N을 가지고 영속상태를 이어가는데 MemberMissions에서 관리되면 되지 않을까요?

sosow0212 commented on src/main/java/com/atwoz/mission/application/membermission/MemberMissionsQueryService.java:
```diff
@@ -1,30 +1,29 @@
+    private final MemberMissionRepository memberMissionRepository;
 
+    public MemberMissionPagingResponse findMemberMissionsWithPaging(final Long memberId, final Pageable pageable) {
+        Page<MemberMissionSimpleResponse> response = memberMissionRepository.findMemberMissionsWithPaging(memberId, pageable);
```
MemberMissions를 통해 조회를 해야할 것 같습니다

sosow0212 commented on src/main/java/com/atwoz/mission/ui/membermission/MemberMissionsController.java:
```diff
@@ -0,0 +1,76 @@
+        return ResponseEntity.ok(memberMissionsQueryService.findMemberMissionsByStatus(memberId, isStatusClear));
+    }
+
+    @GetMapping("/all")
```
해당 로직이 테이블의 값을 조회만 하나요? 이건 GET이 나을까요 POST가 나을까요?

sosow0212 commented on src/main/java/com/atwoz/mission/ui/membermission/MemberMissionsController.java:
```diff
@@ -0,0 +1,76 @@
+                .build();
+    }
+
+    @GetMapping("/{missionId}")
```
이것도요

sosow0212 commented on src/test/java/com/atwoz/mission/ui/membermission/MemberMissionsControllerAcceptanceTest.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    @Nested
+    class 회원_미션_예외 {
```
꼼꼼한 검증 굿~

sosow0212 reviewed (changes requested):
몇 부분만 확인해주시고 반영해주시면 될 것 같습니다~

devholic22 committed:
refactor: 예외 핸들러 HttpStatus 추가

- 예외 핸들러에 HttpStatus를 추가하여 핸들러 함수 최소화 적용
- e 대신 exception으로 명시하도록 수정

devholic22 committed:
refactor: AuthConfig ANY 삭제

- ANY 대신 사용되는 메서드만 표현하도록 수정

devholic22 committed:
docs: index.adoc 추가

- API를 전체 조회하는 index.adoc 파일 추가

devholic22 committed:
refactor: MemberMissionNotFoundException 메시지 수정

- MemberMissionNotFoundException 예외 메시지 수정

devholic22 committed:
refactor: MemberMission mission fetch 타입 변경

- MemberMission의 mission fetch 타입을 LAZY로 변경

devholic22 committed:
refactor: 컨트롤러 API 경로 수정

- 컨트롤러 API 경로 수정

devholic22 committed:
refactor: Restdocs API 경로 수정

- Restdocs API 경로 수정

devholic22 committed:
docs: MemberMissions API 경로 문서 수정

- MemberMissions API 경로 문서 수정

devholic22 committed:
refactor: 보상 로직 수정

- 보상을 조회 (get)에서 수령 (receive)로 변경
- 이미 보상을 수령했을 시 예외 발생되도록 변경
- 테스트 코드 오류 수정

devholic22 committed:
refactor: 보상 로직 추가 검증

- 보상 로직 추가 검증 진행

devholic22 committed:
refactor: 없는 미션 보상 수령 예외 수정

- 미션 조회 실패 예외에서 회원 미션 조회 실패 예외로 변경

devholic22 committed:
docs: MemberMissions API 내용 수정

- MemberMissions API 내용 수정

devholic22 committed:
refactor: 불필요한 API 삭제 처리

- 불필요한 API 삭제 처리
- 관련된 테스트 코드 삭제

devholic22 committed:
refactor: MemberMissionsController 의존성 제거

- MemberMissionsController의 MemberMissionsService 의존성 제거

devholic22 committed:
refactor: 회원 미션 이벤트 방식으로 수정

- 회원 미션 등록, 관리가 이벤트 기반으로 되도록 수정
- 완료한 미션의 미수령 보상 수령 메서드 삭제
- 클리어 상태 삭제 (클리어 된 미션만 들어올 수 있음)
- 성별 별로 가능한 최대 데일리 미션 갯수 반영
- 챌린지 미션은 1회만 등록되도록 설정
- 테스트 코드 예외 수정

devholic22 committed:
refactor: validateDailyMission 메서드 수정

- 메서드명 수정
- 검증 조건을 단축하기 위해 Map으로 초기화

devholic22 committed:
test: memberMission 도메인 테스트 예외 추가 작성

- 성별에 따른 데일리 미션 추가 예외 테스트 작성
- 챌린지 미션 예외 테스트 작성
- 관련 Fixture 등록
- 빌더 사용 시 BaseEntity 사용하도록 SuperBuilder 등록

devholic22 committed:
test: memberMission 서비스 테스트 추가 작성

- 바뀐 도메인에 대한 서비스 테스트 추가 작성
- MethodSource를 활용하여 중복 테스트 공통 처리 진행

devholic22 committed:
refactor: MemberMissions 도메인 테스트 중복 삭제

- MethodSource를 사용하여 비슷한 테스트 삭제 진행

devholic22 committed:
refactor: MemberMission 클리어 여부 조회 삭제

- MemberMission 클리어 여부 조회 기능 삭제
- 관련 테스트 삭제
- 기타 필요없는 데이터 삭제

devholic22 committed:
refactor: MemberMissionsQueryService 메서드 삭제

- 클리어 여부에 따른 회원 미션 조회 메서드 삭제

devholic22 committed:
refactor: 회원 미션 목록 페이징 조회 테스트 수정

- 여러 데일리 미션이 등록되어 있도록 수정

devholic22 committed:
refactor: MemberMission 조회 테스트 수정

- MemberMission 조회 DTO에 계산 로직이 들어가지 않도록 수정
- MemberMission 조회 DTO에 미션 타입, 생성 날짜 들어가도록 수정
- 페이지 처리 검증을 위해 사이즈 수정

devholic22 committed:
docs: 회원 클리어 여부 미션 조회 API 문서 삭제

- 회원 클리어 여부 미션 조회 API 문서 삭제

devholic22 committed:
refactor: MemberMission-Mission 관계 수정

- MemberMission-Mission 관계를 ManyToOne으로 수정

devholic22 committed:
refactor: 미션 추가 검증 로직 조기 리턴 반영

- 넘겨진 미션의 타입에 따라 조기 리턴되도록 검증 메서드 수정

devholic22 committed:
refactor: MemberMissionFixture 메서드명 수정

- MemberMissionFixture 메서드명 수정

devholic22 committed:
refactor: MemberMission 통합 조회 테스트 수정

- MemberMission 통합 조회 테스트 데이터 수정

devholic22 committed:
fix: MemberMissionsService 테스트 메서드 인자 수정

- 미션의 id가 들어가도록 수정

eom-tae-in commented on src/main/java/com/atwoz/mission/application/membermission/MemberMissionsQueryService.java:
```diff
@@ -10,21 +14,22 @@
-        if (isStatusClear) {
-            // TODO
-            return;
+    private static int getNextPage(final int pageNumber, final Page<MemberMissionSimpleResponse> memberMissions) {
```
static 제거해주세요~

devholic22 replied:
제거 완료하였습니다 :)

eom-tae-in commented on src/test/java/com/atwoz/mission/intrastructure/membermission/MemberMissionsFakeRepository.java:
```diff
@@ -0,0 +1,35 @@
+
+    private final Map<Long, MemberMissions> map = new HashMap<>();
+
+    private Long id = 0L;
+
+    @Override
+    public MemberMissions save(final MemberMissions memberMissions) {
+        id++;
```
저장되는 id의 값을 1부터 하고 싶어서 이렇게 하신건가요??
그럼 private Long id = 1L;로 하는게 더 나아 보입니다

devholic22 replied:
수정하였습니다!

eom-tae-in reviewed (commented):
고생하셨습니다! 몇가지만 수정해주세요:)

sosow0212 commented on src/main/java/com/atwoz/mission/application/membermission/MemberMissionsEventHandler.java:
```diff
@@ -0,0 +1,18 @@
+    private final MemberMissionsService memberMissionsService;
+
+    @EventListener
+    public void addClearedMemberMission(final MemberMissionClearedEvent event) {
```
addedClearMemberMission이 맞을 것 같네요!

devholic22 replied:
clear에만 과거형을 붙였었네요 수정 완료하였습니다!

sosow0212 commented on src/main/java/com/atwoz/mission/domain/membermission/MemberMission.java:
```diff
@@ -31,40 +31,38 @@ public class MemberMission extends BaseEntity {
     private Long id;
 
-    @OneToOne(fetch = FetchType.EAGER)
+    @ManyToOne(fetch = FetchType.LAZY)
```
다른 애그리거트인데 해당 방법으로 하신 이유가 있나요
예외가 목적이라면 이벤트로 검증해도 되지 않을까요?

devholic22 replied:
MemberMission에서 보상을 수령할 때 미션으로부터 받게 되고, 회원이 가지고 있는 미션이 챌린지 미션인지를 판단하는 과정에서도 미션으로부터 메서드를 호출하게 됩니다.
또한, MemberMission이 미션에 대한 id만 가지고 있다면 MemberMission을 조회할 때 제약이 너무 많이 생길 것 같다는 생각이 들었습니다. (챌린지/데일리 여부, 미션의 보상 등등)
그래서 서로 다른 애그리거트더라도, 이 경우에 대해서는 객체 참조로 하는 게 적절할 것 같다는 판단을 하였습니다.

이벤트 방식으로 하면 `이벤트를 수신한 객체로부터 값을 전달받는 것`에 대해 제약이 많을 것 같은데, 재윤님께서는 이러한 상황에서 어떻게 해결하시는지 여쭤봐도 될까요?

sosow0212 replied:
음 이 부분은 생성할 때 역정규화로 값을 저장해도 될 것 같아요
조금 어색하게 느껴지는데 일단 현준님 말도 충분히 이해는 돼서 개발하다가 문제가 생기면 그때 바꾸는 걸로 하도록 합시다!

devholic22 committed:
refactor: 메서드 static 제거

- MemberMissionsQueryService static 제거

sosow0212 commented on src/main/java/com/atwoz/mission/domain/membermission/MemberMissions.java:
```diff
@@ -38,22 +53,67 @@ public class MemberMissions extends BaseEntity {
     }
 
-    public Integer getRewardBy(final Long missionId) {
+    private void validateIsCanAddMission(final Gender memberGender, final Mission mission) {
```
isCan은 네이밍이 조금 이상한 것 같아요
validateCan~~이 어떨까요?

devholic22 replied:
수정하였습니다! is can 둘 다 동사였는데 함께 사용했었네요

sosow0212 commented on src/main/java/com/atwoz/mission/domain/membermission/MemberMissions.java:
```diff
@@ -38,22 +53,67 @@ public class MemberMissions extends BaseEntity {
+            return;
+        }
+
+        List<MemberMission> existedMission = extractChallengeMission(mission);
+
+        if (!existedMission.isEmpty()) {
+            throw new AlreadyChallengeMissionExistedException();
+        }
```
이 부분도 extractChallengeMission 메서드에서 스트림으로 한 번에 할 수 있어보이네요

devholic22 replied:
ifPresent를 이용하여 수정하였습니다!

sosow0212 commented on src/main/java/com/atwoz/mission/intrastructure/membermission/MemberMissionQueryRepository.java:
```diff
@@ -0,0 +1,41 @@
+
+    private final JPAQueryFactory jpaQueryFactory;
+
+    public Page<MemberMissionSimpleResponse> findMemberMissionsWithPaging(final Long memberId, final Pageable pageable) {
```
해당 쿼리는 어떻게 나갈까요?

devholic22 committed:
refactor: 생성 날짜 비교 메서드화

- 데일리 미션 생성 날짜 비교 메서드 추가

sosow0212 commented on src/main/java/com/atwoz/mission/application/membermission/MemberMissionsService.java:
```diff
@@ -9,24 +16,37 @@
-
+    
     private final MemberMissionsRepository memberMissionsRepository;
+    private final MissionRepository missionRepository;
```
MemberMission 쪽이 모두 Mission에 영향을 받을 거 같은데 Mission 참조는 없어도 될 것 같습니다 이벤트나 다른 걸로 풀어나가면 어떨까요?

sosow0212 reviewed (changes requested):
수고하셨습니다 몇 가지만 확인하고 수정해주시면 될 것 같습니다~

devholic22 committed:
refactor: FakeRepository id 초깃값 수정

- MemberMissionsFakeRepository id 초깃값 수정

devholic22 committed:
refactor: addClearedMemberMission 시제 변경

- 이벤트 컨벤션에 맞게 과거형으로 네이밍

devholic22 committed:
refactor: MemberMissions 메서드명 수정

- 동사 중복 삭제 (IsCan -> Can)

devholic22 committed:
refactor: MemberMission 검증 메서드 수정

- 데일리 미션에서 스트림을 활용하여 한 번에 표현되도록 수정
- extractTodayDailyMissions 반환 타입 수정

devholic22 committed:
refactor: MemberMissionRepository 삭제

- MemberMissionRepository 삭제
- MemberMissionQueryRepository 삭제
- MemberMissionsRepository만 존재하도록 수정

sosow0212 reviewed (approved):
작업하시느라 수고하셨습니다!
